frontend/include/chpl/resolution/resolution-types.h

```diff
@@ -885,6 +885,103 @@ class TypedFnSignature {
   /// \endcond DO_NOT_DOCUMENT
 };
 
+// Container for resolution candidates and (if applicable) their correponding
+// forwarding-to types.
+struct CandidatesAndForwardingInfo {
+  private:
+  std::vector<const TypedFnSignature*> candidates;
+  std::vector<types::QualifiedType> forwardingInfo;
+
+  public:
+   // Add a candidate and optional forwarding info.
+   void addCandidate(const TypedFnSignature* candidate,
+                     const types::QualifiedType* forwardingTo = nullptr) {
+     candidates.push_back(candidate);
+     if (forwardingTo) {
+       forwardingInfo.push_back(*forwardingTo);
+     }
+   }
+
+   // Compute and fill in forwarding info for a range of newly-added candidates.
+   void helpComputeForwardingTo(const CallInfo& fci, size_t start) {
+     types::QualifiedType forwardingReceiverActualType = fci.calledType();
+     forwardingInfo.resize(start);
```

Comment:
This could silently drop existing forwarding infos. Shouldn't you assert that `forwardingInfo.size() <= start`? I see this code is extracted, but perhaps it's good to add the assertion anyway.